File: ai_v4/lint-relations.js
#!/usr/bin/env node
/**
 * Linter relacji między dokumentami.
 *
 * Czyta katalog z plikami .md (front matter YAML + treść), buduje graf
 * z pola `relations` i sprawdza go względem identyfikatorów użytych w treści.
 *
 * Stare pola `source-spec-ids` i `related-doc-ids` są ignorowane.
 *
 * Użycie:
 *     node lint-relations.js <katalog> [--strict]
 *
 * Zależności: js-yaml  (npm install js-yaml)
 * Wynik: raport tekstowy na stdout. Kod wyjścia 1, gdy są błędy.
 */

"use strict";

const fs = require("fs");
const path = require("path");

let yaml;
try {
  yaml = require("js-yaml");
} catch {
  console.error("Brak modułu js-yaml. Zainstaluj: npm install js-yaml");
  process.exit(2);
}

// ---------------------------------------------------------------------------
// Słownik standardu
// ---------------------------------------------------------------------------

const RELATION_TYPES = new Set([
  "defines", "realizes", "constrained-by", "uses-data", "consumes",
  "involves", "used-in", "contains", "groups", "precedes", "supersedes",
]);

// Prefiks doc-id → skrót typu (dłuższe pierwsze, żeby SPEC-WF nie złapało się jako SPEC)
const DOC_PREFIXES = [
  "SPEC-WF", "SCRSEC", "UCMAP", "ACTOR", "BPMN", "FLOW", "TUC", "DDM",
  "LDM", "ENT", "CAP", "NFR", "API", "INT", "QUE", "ACT", "STM", "SCR",
  "NAV", "ADR", "BFS", "UC",
];

const TYPE_FIELD_TO_PREFIX = {
  "business-functional-spec": "BFS",
  "use-case": "UC",
  "technical-use-case": "TUC",
  "use-case-technical-map": "UCMAP",
  "actor": "ACTOR",
  "ddm": "DDM",
  "ldm": "LDM",
  "ldm-entity": "ENT",
  "capability": "CAP",
  "non-functional-requirements": "NFR",
  "api": "API",
  "integration": "INT",
  "event": "QUE",
  "flow": "FLOW",
  "activity": "ACT",
  "state-machine": "STM",
  "screen": "SCR",
  "screen-section": "SCRSEC",
  "navigation-map": "NAV",
  "bpmn-process": "BPMN",
  "workflow-step": "SPEC-WF",
  "adr": "ADR",
};

// Identyfikatory lokalne: który typ dokumentu je definiuje i jaki mają wzorzec
const LOCAL_ID_OWNERS = {
  "BFS": [/REQ-\d+/g, /RB-\d+/g],
  "DDM": [/DOM-[A-Za-z][\w-]*/g, /RD-\d+/g],
  "LDM": [/RW-\d+/g],
  "NFR": [/NFR-[A-Z]+-\d+/g],
  "ACTOR": [/ACTOR-\d{1,2}(?!\d)/g],
  "SPEC-WF": [/AC-\d+/g],
};
// Warianty A1/E1 — tylko z nagłówków
const VARIANT_HEADING = /^#{2,4}\s+([AE]\d+)\s*[:.]/gm;

// Sugerowany typ relacji dla pary (typ źródła → typ celu)
const PAIR_TO_TYPE = new Map([
  ["BFS→NFR", "constrained-by"],
  ["UC→BFS", "realizes"], ["TUC→BFS", "realizes"],
  ["UC→CAP", "realizes"], ["TUC→CAP", "realizes"],
  ["UC→ACTOR", "involves"], ["TUC→ACTOR", "involves"],
  ["CAP→BFS", "realizes"],
  ["CAP→QUE", "consumes"],
  ["API→CAP", "realizes"], ["INT→CAP", "realizes"], ["QUE→CAP", "realizes"],
  ["API→UC", "used-in"], ["API→TUC", "used-in"],
  ["INT→UC", "used-in"], ["INT→TUC", "used-in"],
  ["QUE→UC", "used-in"], ["QUE→TUC", "used-in"],
  ["API→ENT", "uses-data"], ["SCRSEC→ENT", "uses-data"],
  ["LDM→DDM", "realizes"],
  ["LDM→ENT", "contains"],
  ["ENT→DDM", "realizes"],
  ["ENT→ENT", "contains"],
  ["STM→DDM", "realizes"],
  ["SCR→SCRSEC", "contains"],
  ["SCR→UC", "used-in"], ["SCR→TUC", "used-in"],
  ["NAV→SCR", "groups"],
  ["UCMAP→UC", "groups"], ["UCMAP→TUC", "groups"],
  ["FLOW→ACT", "contains"],
  ["FLOW→API", "consumes"], ["FLOW→INT", "consumes"], ["FLOW→QUE", "consumes"],
  ["BPMN→SPEC-WF", "contains"],
  ["SPEC-WF→SPEC-WF", "precedes"],
  ["SPEC-WF→API", "consumes"], ["SPEC-WF→INT", "consumes"], ["SPEC-WF→QUE", "consumes"],
  ["ADR→ADR", "supersedes"],
]);

function suggestType(src, dst) {
  const key = `${src}→${dst}`;
  if (PAIR_TO_TYPE.has(key)) return PAIR_TO_TYPE.get(key);
  if (dst === "NFR" || dst === "ADR") return "constrained-by";
  return null;
}

// ---------------------------------------------------------------------------
// Parsowanie
// ---------------------------------------------------------------------------

const FRONT_MATTER = /^---\s*\n([\s\S]*?)\n---\s*\n?([\s\S]*)$/;
const HTML_COMMENT = /<!--[\s\S]*?-->/g;
const CODE_FENCE = /```[\s\S]*?```/g;

const esc = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
const PREFIX_ALT = DOC_PREFIXES.map(esc).join("|");
const DOC_ID_RE = new RegExp(`\\b(${PREFIX_ALT})-(\\d{3,}|[A-Z][A-Za-z]+)\\b`, "g");
const PREFIXED_LOCAL_RE = new RegExp(
  `\\b((?:${PREFIX_ALT})-(?:\\d{3,}|[A-Z][A-Za-z]+))\\.([A-Z]+-[\\w-]+)\\b`, "g"
);

function matchAll(re, text) {
  re.lastIndex = 0;
  return [...text.matchAll(re)];
}

class Doc {
  constructor(file, meta, body) {
    this.path = file;
    this.meta = meta || {};
    this.body = body;
    this.docId = String(this.meta["doc-id"] || "").trim();
    this.kind = this.resolveKind();
    this.badRelations = false;
    this.badEntries = [];
    this.relations = this.parseRelations();
    this.errors = [];
    this.warnings = [];
    this.infos = [];
  }

  resolveKind() {
    const t = this.meta.type;
    if (t && TYPE_FIELD_TO_PREFIX[t]) return TYPE_FIELD_TO_PREFIX[t];
    for (const p of DOC_PREFIXES) {
      if (this.docId.startsWith(p + "-")) return p;
    }
    return null;
  }

  parseRelations() {
    const raw = this.meta.relations || [];
    if (!Array.isArray(raw)) {
      this.badRelations = true;
      return [];
    }
    const out = [];
    for (const r of raw) {
      if (r && typeof r === "object" && "type" in r && "target" in r) {
        out.push({ type: String(r.type).trim(), target: String(r.target).trim() });
      } else {
        this.badEntries.push(r);
      }
    }
    return out;
  }

  cleanBody() {
    return this.body.replace(HTML_COMMENT, "").replace(CODE_FENCE, "");
  }
}

function walk(dir, out = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) walk(full, out);
    else if (entry.isFile() && entry.name.endsWith(".md") && !entry.name.startsWith("_")) {
      out.push(full);
    }
  }
  return out.sort();
}

function loadDocs(root) {
  const docs = new Map();
  const skipped = [];
  for (const file of walk(root)) {
    const text = fs.readFileSync(file, "utf8");
    const m = FRONT_MATTER.exec(text);
    if (!m) {
      skipped.push([file, "brak front matter"]);
      continue;
    }
    let meta;
    try {
      meta = yaml.load(m[1]) || {};
    } catch (e) {
      skipped.push([file, `błąd YAML: ${e.message}`]);
      continue;
    }
    const d = new Doc(file, meta, m[2]);
    if (!d.docId) {
      skipped.push([file, "brak doc-id"]);
      continue;
    }
    if (d.docId.startsWith("TPL-")) continue;
    if (docs.has(d.docId)) {
      d.errors.push(`duplikat doc-id \`${d.docId}\` — także w ${docs.get(d.docId).path}`);
    }
    docs.set(d.docId, d);
  }
  return { docs, skipped };
}

// ---------------------------------------------------------------------------
// Reguły
// ---------------------------------------------------------------------------

function kindOfTarget(target, docs) {
  const dot = target.indexOf(".");
  if (dot !== -1) {
    const docPart = target.slice(0, dot);
    const local = target.slice(dot + 1);
    const d = docs.get(docPart);
    return { kind: d ? d.kind : null, docId: docPart, local };
  }
  const d = docs.get(target);
  return { kind: d ? d.kind : null, docId: target, local: null };
}

function checkDocument(doc, docs, allDefined) {
  const body = doc.cleanBody();
  const declared = new Set(doc.relations.map((r) => r.target));

  // R1: poprawność wpisów relations
  if (doc.badRelations) doc.errors.push("pole `relations` nie jest listą");
  for (const bad of doc.badEntries) {
    doc.errors.push(`wpis w \`relations\` bez \`type\` lub \`target\`: ${JSON.stringify(bad)}`);
  }
  for (const { type, target } of doc.relations) {
    if (!RELATION_TYPES.has(type)) {
      doc.errors.push(`nieznany typ relacji \`${type}\` → ${target}`);
    }
  }

  // R2: cel relacji istnieje
  for (const { type, target } of doc.relations) {
    if (type === "defines") continue;
    const { docId, local } = kindOfTarget(target, docs);
    if (!docs.has(docId)) {
      doc.errors.push(`cel nie istnieje: \`${target}\` (relacja \`${type}\`)`);
    } else if (local && !allDefined.has(target)) {
      doc.errors.push(`cel \`${target}\` nie jest zadeklarowany przez \`defines\` w ${docId}`);
    }
  }

  // R3: własne identyfikatory lokalne muszą mieć `defines`
  const ownLocal = new Set();
  for (const re of LOCAL_ID_OWNERS[doc.kind] || []) {
    for (const m of matchAll(re, body)) ownLocal.add(m[0]);
  }
  if (doc.kind === "UC" || doc.kind === "TUC") {
    for (const m of matchAll(VARIANT_HEADING, body)) ownLocal.add(m[1]);
  }
  const definedHere = new Set(doc.relations.filter((r) => r.type === "defines").map((r) => r.target));
  for (const lid of [...ownLocal].filter((x) => !definedHere.has(x)).sort()) {
    doc.errors.push(`identyfikator \`${lid}\` w treści bez wpisu \`defines\``);
  }
  for (const lid of [...definedHere].filter((x) => !ownLocal.has(x)).sort()) {
    doc.warnings.push(`\`defines: ${lid}\` bez śladu w treści`);
  }

  // R4: identyfikatory innych dokumentów w treści muszą mieć relację
  const mentioned = new Set();
  const prefixedSpans = [];
  for (const m of matchAll(PREFIXED_LOCAL_RE, body)) {
    mentioned.add(m[0]);
    prefixedSpans.push([m.index, m.index + m[0].length]);
  }
  for (const m of matchAll(DOC_ID_RE, body)) {
    const did = m[0];
    if (prefixedSpans.some(([a, b]) => a <= m.index && m.index < b)) continue;
    if (did !== doc.docId && docs.has(did)) mentioned.add(did);
  }
  const declaredDocs = new Set([...declared].map((t) => t.split(".")[0]));
  for (const target of [...mentioned].sort()) {
    if (declared.has(target)) continue;
    if (!target.includes(".") && declaredDocs.has(target)) continue;
    const { kind: tkind, docId } = kindOfTarget(target, docs);
    if (!docs.has(docId)) {
      doc.errors.push(`w treści \`${target}\`, ale taki dokument nie istnieje`);
      continue;
    }
    const stype = suggestType(doc.kind, tkind);
    const hint = stype
      ? `dodaj \`- type: ${stype}\` / \`target: ${target}\``
      : `standard nie przewiduje relacji ${doc.kind} → ${tkind}; usuń odwołanie z treści albo zadeklaruj je po stronie ${tkind}`;
    doc.errors.push(`w treści \`${target}\` bez relacji — ${hint}`);
  }

  // R5: identyfikatory lokalne bez prefiksu — stan „w trakcie"
  const unprefixed = new Set();
  for (const [kind, patterns] of Object.entries(LOCAL_ID_OWNERS)) {
    if (kind === doc.kind) continue;
    for (const re of patterns) {
      for (const m of matchAll(re, body)) {
        if (m.index > 0 && body[m.index - 1] === ".") continue;
        unprefixed.add(m[0]);
      }
    }
  }
  for (const lid of [...unprefixed].sort()) {
    doc.infos.push(`w trakcie: \`${lid}\` bez prefiksu dokumentu — nie da się rozwiązać`);
  }

  // R6: pary spoza macierzy
  for (const { type, target } of doc.relations) {
    if (type === "defines" || !RELATION_TYPES.has(type)) continue;
    const { kind: tkind } = kindOfTarget(target, docs);
    const expected = tkind ? suggestType(doc.kind, tkind) : null;
    if (expected && expected !== type) {
      doc.warnings.push(
        `\`${type}\` → ${target}: dla pary ${doc.kind} → ${tkind} standard przewiduje \`${expected}\``
      );
    }
  }
}

function checkConnectivity(docs) {
  const adj = new Map();
  const nodes = new Set(docs.keys());
  const link = (a, b) => {
    if (!adj.has(a)) adj.set(a, new Set());
    if (!adj.has(b)) adj.set(b, new Set());
    adj.get(a).add(b);
    adj.get(b).add(a);
  };
  for (const d of docs.values()) {
    for (const { type, target } of d.relations) {
      const node = type === "defines" ? `${d.docId}.${target}` : target;
      nodes.add(node);
      link(d.docId, node);
    }
  }
  const seen = new Set();
  const components = [];
  for (const n of [...nodes].sort()) {
    if (seen.has(n)) continue;
    const comp = new Set();
    const stack = [n];
    while (stack.length) {
      const x = stack.pop();
      if (comp.has(x)) continue;
      comp.add(x);
      for (const y of adj.get(x) || []) if (!comp.has(y)) stack.push(y);
    }
    for (const x of comp) seen.add(x);
    components.push(comp);
  }
  components.sort((a, b) => b.size - a.size);
  return components;
}

// ---------------------------------------------------------------------------
// Raport
// ---------------------------------------------------------------------------

function main() {
  const args = process.argv.slice(2);
  const root = args.find((a) => !a.startsWith("--"));
  const strict = args.includes("--strict");
  if (!root) {
    console.error("Użycie: node lint-relations.js <katalog> [--strict]");
    process.exit(2);
  }

  const { docs, skipped } = loadDocs(root);
  if (docs.size === 0) {
    console.error("Nie znaleziono dokumentów z doc-id.");
    process.exit(2);
  }

  const allDefined = new Set();
  for (const d of docs.values()) {
    for (const { type, target } of d.relations) {
      if (type === "defines") allDefined.add(`${d.docId}.${target}`);
    }
  }

  for (const d of docs.values()) checkDocument(d, docs, allDefined);

  let nErr = 0, nWarn = 0, nInfo = 0;
  for (const did of [...docs.keys()].sort()) {
    const d = docs.get(did);
    if (!d.errors.length && !d.warnings.length && !d.infos.length) continue;
    console.log(`\n${did}  (${d.path})`);
    for (const e of d.errors) console.log(`  BŁĄD   ${e}`);
    for (const w of d.warnings) console.log(`  UWAGA  ${w}`);
    for (const i of d.infos) console.log(`  INFO   ${i}`);
    nErr += d.errors.length;
    nWarn += d.warnings.length;
    nInfo += d.infos.length;
  }

  const components = checkConnectivity(docs);
  if (components.length > 1) {
    console.log("\nSPÓJNOŚĆ (U-003)");
    console.log(`  główna składowa: ${components[0].size} węzłów`);
    const islands = components.slice(1);
    console.log(`  poza nią: ${islands.length} składowych`);
    for (const comp of islands.slice(0, 20)) {
      const ids = [...comp].filter((x) => docs.has(x)).sort();
      console.log(`    ${ids.length ? ids.join(", ") : "(tylko identyfikatory lokalne)"}`);
    }
    if (islands.length > 20) console.log(`    … i ${islands.length - 20} więcej`);
    if (strict) nErr += islands.length;
  }

  if (skipped.length) {
    console.log("\nPOMINIĘTE");
    for (const [p, why] of skipped) console.log(`  ${p}: ${why}`);
  }

  console.log(`\n${docs.size} dokumentów · ${nErr} błędów · ${nWarn} uwag · ${nInfo} w trakcie`);
  process.exit(nErr ? 1 : 0);
}

main();

File: ai_v4/lint_relations.py
#!/usr/bin/env python3
"""
Linter relacji między dokumentami.

Czyta katalog z plikami .md (front matter YAML + treść), buduje graf
z pola `relations` i sprawdza go względem identyfikatorów użytych w treści.

Stare pola `source-spec-ids` i `related-doc-ids` są ignorowane.

Użycie:
    python lint_relations.py <katalog> [--strict]

Wynik: raport tekstowy na stdout. Kod wyjścia 1, gdy są błędy.
"""

import re
import sys
from collections import defaultdict
from pathlib import Path

try:
    import yaml
except ImportError:
    sys.exit("Brak modułu PyYAML. Zainstaluj: pip install pyyaml")

# ---------------------------------------------------------------------------
# Słownik standardu
# ---------------------------------------------------------------------------

RELATION_TYPES = {
    "defines", "realizes", "constrained-by", "uses-data", "consumes",
    "involves", "used-in", "contains", "groups", "precedes", "supersedes",
}

# Prefiks doc-id → skrót typu
DOC_PREFIXES = [
    "SPEC-WF", "SCRSEC", "UCMAP", "ACTOR", "BPMN", "FLOW", "TUC", "DDM",
    "LDM", "ENT", "CAP", "NFR", "API", "INT", "QUE", "ACT", "STM", "SCR",
    "NAV", "ADR", "BFS", "UC",
]

# Pole `type` z front matter → skrót typu
TYPE_FIELD_TO_PREFIX = {
    "business-functional-spec": "BFS",
    "use-case": "UC",
    "technical-use-case": "TUC",
    "use-case-technical-map": "UCMAP",
    "actor": "ACTOR",
    "ddm": "DDM",
    "ldm": "LDM",
    "ldm-entity": "ENT",
    "capability": "CAP",
    "non-functional-requirements": "NFR",
    "api": "API",
    "integration": "INT",
    "event": "QUE",
    "flow": "FLOW",
    "activity": "ACT",
    "state-machine": "STM",
    "screen": "SCR",
    "screen-section": "SCRSEC",
    "navigation-map": "NAV",
    "bpmn-process": "BPMN",
    "workflow-step": "SPEC-WF",
    "adr": "ADR",
}

# Identyfikatory lokalne: który typ dokumentu je definiuje i jaki mają wzorzec
LOCAL_ID_OWNERS = {
    "BFS": [r"REQ-\d+", r"RB-\d+"],
    "DDM": [r"DOM-[A-Za-z][\w-]*", r"RD-\d+"],
    "LDM": [r"RW-\d+"],
    "NFR": [r"NFR-[A-Z]+-\d+"],
    "ACTOR": [r"ACTOR-\d{1,2}(?!\d)"],
    "SPEC-WF": [r"AC-\d+"],
}
# Warianty A1/E1 — tylko z nagłówków, bo same litery są zbyt ogólne
VARIANT_HEADING = re.compile(r"^#{2,4}\s+([AE]\d+)\s*[:.]", re.MULTILINE)

# Sugerowany typ relacji dla pary (typ źródła, typ celu)
PAIR_TO_TYPE = {
    ("BFS", "NFR"): "constrained-by",
    ("UC", "BFS"): "realizes", ("TUC", "BFS"): "realizes",
    ("UC", "CAP"): "realizes", ("TUC", "CAP"): "realizes",
    ("UC", "ACTOR"): "involves", ("TUC", "ACTOR"): "involves",
    ("CAP", "BFS"): "realizes",
    ("CAP", "QUE"): "consumes",
    ("API", "CAP"): "realizes", ("INT", "CAP"): "realizes", ("QUE", "CAP"): "realizes",
    ("API", "UC"): "used-in", ("API", "TUC"): "used-in",
    ("INT", "UC"): "used-in", ("INT", "TUC"): "used-in",
    ("QUE", "UC"): "used-in", ("QUE", "TUC"): "used-in",
    ("API", "ENT"): "uses-data", ("SCRSEC", "ENT"): "uses-data",
    ("LDM", "DDM"): "realizes",
    ("LDM", "ENT"): "contains",
    ("ENT", "DDM"): "realizes",
    ("ENT", "ENT"): "contains",
    ("STM", "DDM"): "realizes",
    ("SCR", "SCRSEC"): "contains",
    ("SCR", "UC"): "used-in", ("SCR", "TUC"): "used-in",
    ("NAV", "SCR"): "groups",
    ("UCMAP", "UC"): "groups", ("UCMAP", "TUC"): "groups",
    ("FLOW", "ACT"): "contains",
    ("FLOW", "API"): "consumes", ("FLOW", "INT"): "consumes", ("FLOW", "QUE"): "consumes",
    ("BPMN", "SPEC-WF"): "contains",
    ("SPEC-WF", "SPEC-WF"): "precedes",
    ("SPEC-WF", "API"): "consumes", ("SPEC-WF", "INT"): "consumes", ("SPEC-WF", "QUE"): "consumes",
    ("ADR", "ADR"): "supersedes",
}


def suggest_type(src, dst):
    if (src, dst) in PAIR_TO_TYPE:
        return PAIR_TO_TYPE[(src, dst)]
    if dst == "NFR" or dst == "ADR":
        return "constrained-by"
    return None


# ---------------------------------------------------------------------------
# Parsowanie
# ---------------------------------------------------------------------------

FRONT_MATTER = re.compile(r"\A---\s*\n(.*?)\n---\s*\n?(.*)\Z", re.DOTALL)
HTML_COMMENT = re.compile(r"<!--.*?-->", re.DOTALL)
CODE_FENCE = re.compile(r"```.*?```", re.DOTALL)

DOC_ID_RE = re.compile(
    r"\b(" + "|".join(re.escape(p) for p in DOC_PREFIXES) + r")-(\d{3,}|[A-Z][A-Za-z]+)\b"
)
PREFIXED_LOCAL_RE = re.compile(
    r"\b((?:" + "|".join(re.escape(p) for p in DOC_PREFIXES) + r")-(?:\d{3,}|[A-Z][A-Za-z]+))"
    r"\.([A-Z]+-[\w-]+)\b"
)


class Doc:
    def __init__(self, path, meta, body):
        self.path = path
        self.meta = meta or {}
        self.body = body
        self.doc_id = str(self.meta.get("doc-id", "")).strip()
        self.kind = self._resolve_kind()
        self.relations = self._parse_relations()
        self.errors, self.warnings, self.infos = [], [], []

    def _resolve_kind(self):
        t = self.meta.get("type")
        if t in TYPE_FIELD_TO_PREFIX:
            return TYPE_FIELD_TO_PREFIX[t]
        for p in DOC_PREFIXES:
            if self.doc_id.startswith(p + "-"):
                return p
        return None

    def _parse_relations(self):
        out = []
        raw = self.meta.get("relations") or []
        if not isinstance(raw, list):
            self.meta["_bad_relations"] = True
            return out
        for r in raw:
            if isinstance(r, dict) and "type" in r and "target" in r:
                out.append((str(r["type"]).strip(), str(r["target"]).strip()))
            else:
                self.meta.setdefault("_bad_entries", []).append(r)
        return out

    def clean_body(self):
        b = HTML_COMMENT.sub("", self.body)
        b = CODE_FENCE.sub("", b)
        return b


def load_docs(root):
    docs = {}
    skipped = []
    for p in sorted(Path(root).rglob("*.md")):
        if p.name.startswith("_"):
            continue
        text = p.read_text(encoding="utf-8", errors="replace")
        m = FRONT_MATTER.match(text)
        if not m:
            skipped.append((p, "brak front matter"))
            continue
        try:
            meta = yaml.safe_load(m.group(1)) or {}
        except yaml.YAMLError as e:
            skipped.append((p, f"błąd YAML: {e}"))
            continue
        d = Doc(p, meta, m.group(2))
        if not d.doc_id:
            skipped.append((p, "brak doc-id"))
            continue
        if d.doc_id.startswith("TPL-"):
            continue
        if d.doc_id in docs:
            d.errors.append(f"duplikat doc-id `{d.doc_id}` — także w {docs[d.doc_id].path}")
        docs[d.doc_id] = d
    return docs, skipped


# ---------------------------------------------------------------------------
# Reguły
# ---------------------------------------------------------------------------

def kind_of_target(target, docs):
    """Zwraca (skrót typu, doc-id celu, czy identyfikator lokalny)."""
    if "." in target:
        doc_part, local = target.split(".", 1)
        d = docs.get(doc_part)
        return (d.kind if d else None), doc_part, local
    d = docs.get(target)
    return (d.kind if d else None), target, None


def check_document(doc, docs, all_defined):
    body = doc.clean_body()
    declared = {t for _, t in doc.relations}

    # R1: poprawność wpisów relations
    if doc.meta.get("_bad_relations"):
        doc.errors.append("pole `relations` nie jest listą")
    for bad in doc.meta.get("_bad_entries", []):
        doc.errors.append(f"wpis w `relations` bez `type` lub `target`: {bad!r}")
    for rtype, target in doc.relations:
        if rtype not in RELATION_TYPES:
            doc.errors.append(f"nieznany typ relacji `{rtype}` → {target}")

    # R2: cel relacji istnieje
    for rtype, target in doc.relations:
        if rtype == "defines":
            continue
        tkind, tdoc, local = kind_of_target(target, docs)
        if tdoc not in docs:
            doc.errors.append(f"cel nie istnieje: `{target}` (relacja `{rtype}`)")
        elif local and target not in all_defined:
            doc.errors.append(
                f"cel `{target}` nie jest zadeklarowany przez `defines` w {tdoc}"
            )

    # R3: własne identyfikatory lokalne muszą mieć `defines`
    own_local = set()
    for pat in LOCAL_ID_OWNERS.get(doc.kind, []):
        own_local.update(re.findall(pat, body))
    if doc.kind in ("UC", "TUC"):
        own_local.update(VARIANT_HEADING.findall(body))
    defined_here = {t for r, t in doc.relations if r == "defines"}
    for lid in sorted(own_local - defined_here):
        doc.errors.append(f"identyfikator `{lid}` w treści bez wpisu `defines`")
    for lid in sorted(defined_here - own_local):
        doc.warnings.append(f"`defines: {lid}` bez śladu w treści")

    # R4: identyfikatory innych dokumentów w treści muszą mieć relację
    mentioned = set()
    prefixed_spans = []
    for m in PREFIXED_LOCAL_RE.finditer(body):
        prefixed_spans.append((m.start(), m.end()))
        if m.group(1) == doc.doc_id:
            continue  # własny identyfikator z prefiksem — sprawdza R3
        mentioned.add(m.group(0))
    for m in DOC_ID_RE.finditer(body):
        did = m.group(0)
        if any(a <= m.start() < b for a, b in prefixed_spans):
            continue  # część identyfikatora z prefiksem, obsłużona wyżej
        if did != doc.doc_id and did in docs:
            mentioned.add(did)
    declared_docs = {t.split(".", 1)[0] for t in declared}
    for target in sorted(mentioned):
        if target in declared:
            continue
        if "." not in target and target in declared_docs:
            continue  # relacja do identyfikatora lokalnego z tego dokumentu wystarcza
        tkind, tdoc, local = kind_of_target(target, docs)
        if tdoc not in docs:
            doc.errors.append(f"w treści `{target}`, ale taki dokument nie istnieje")
            continue
        stype = suggest_type(doc.kind, tkind)
        if stype:
            hint = f"dodaj `- type: {stype}` / `target: {target}`"
        else:
            hint = (f"standard nie przewiduje relacji {doc.kind} → {tkind}; "
                    f"usuń odwołanie z treści albo zadeklaruj je po stronie {tkind}")
        doc.errors.append(f"w treści `{target}` bez relacji — {hint}")

    # R5: identyfikatory lokalne bez prefiksu — stan „w trakcie"
    foreign_patterns = [
        pat for k, pats in LOCAL_ID_OWNERS.items() if k != doc.kind for pat in pats
    ]
    unprefixed = set()
    for pat in foreign_patterns:
        for m in re.finditer(pat, body):
            start = m.start()
            if start > 0 and body[start - 1] == ".":
                continue
            unprefixed.add(m.group(0))
    for lid in sorted(unprefixed):
        doc.infos.append(f"w trakcie: `{lid}` bez prefiksu dokumentu — nie da się rozwiązać")

    # R6: pary spoza macierzy
    for rtype, target in doc.relations:
        if rtype == "defines" or rtype not in RELATION_TYPES:
            continue
        tkind, _, _ = kind_of_target(target, docs)
        if tkind and suggest_type(doc.kind, tkind) not in (None, rtype):
            doc.warnings.append(
                f"`{rtype}` → {target}: dla pary {doc.kind} → {tkind} standard "
                f"przewiduje `{suggest_type(doc.kind, tkind)}`"
            )


def check_connectivity(docs):
    """U-003: każdy węzeł osiągalny. Graf nieskierowany z istniejących relacji."""
    adj = defaultdict(set)
    nodes = set(docs)
    for d in docs.values():
        for rtype, target in d.relations:
            node = f"{d.doc_id}.{target}" if rtype == "defines" else target
            nodes.add(node)
            adj[d.doc_id].add(node)
            adj[node].add(d.doc_id)
    seen, components = set(), []
    for n in sorted(nodes):
        if n in seen:
            continue
        comp, stack = set(), [n]
        while stack:
            x = stack.pop()
            if x in comp:
                continue
            comp.add(x)
            stack.extend(adj[x] - comp)
        seen |= comp
        components.append(comp)
    components.sort(key=len, reverse=True)
    return components


# ---------------------------------------------------------------------------
# Raport
# ---------------------------------------------------------------------------

def main():
    if len(sys.argv) < 2:
        sys.exit(__doc__)
    root = sys.argv[1]
    strict = "--strict" in sys.argv

    docs, skipped = load_docs(root)
    if not docs:
        sys.exit("Nie znaleziono dokumentów z doc-id.")

    all_defined = set()
    for d in docs.values():
        for r, t in d.relations:
            if r == "defines":
                all_defined.add(f"{d.doc_id}.{t}")

    for d in docs.values():
        check_document(d, docs, all_defined)

    n_err = n_warn = n_info = 0
    for did in sorted(docs):
        d = docs[did]
        if not (d.errors or d.warnings or d.infos):
            continue
        print(f"\n{did}  ({d.path})")
        for e in d.errors:
            print(f"  BŁĄD   {e}")
        for w in d.warnings:
            print(f"  UWAGA  {w}")
        for i in d.infos:
            print(f"  INFO   {i}")
        n_err += len(d.errors)
        n_warn += len(d.warnings)
        n_info += len(d.infos)

    components = check_connectivity(docs)
    if len(components) > 1:
        print("\nSPÓJNOŚĆ (U-003)")
        print(f"  główna składowa: {len(components[0])} węzłów")
        islands = components[1:]
        print(f"  poza nią: {len(islands)} składowych")
        for comp in islands[:20]:
            ids = sorted(x for x in comp if x in docs)
            print(f"    {', '.join(ids) if ids else '(tylko identyfikatory lokalne)'}")
        if len(islands) > 20:
            print(f"    … i {len(islands) - 20} więcej")
        if strict:
            n_err += len(islands)

    if skipped:
        print("\nPOMINIĘTE")
        for p, why in skipped:
            print(f"  {p}: {why}")

    print(f"\n{len(docs)} dokumentów · {n_err} błędów · {n_warn} uwag · {n_info} w trakcie")
    sys.exit(1 if n_err else 0)


if __name__ == "__main__":
    main()
